File: 프로그래머스/3/43163. 단어 변환/단어 변환.java
import java.util.*;

class Solution {
    public int solution(String begin, String target, String[] words) {
        int answer = 0;
        boolean[] visited = new boolean[words.length];
        Queue<ChangedWord> queue = new LinkedList<>();
        queue.offer(new ChangedWord(begin, 0));
        
        while(!queue.isEmpty()) {
            ChangedWord word = queue.poll();
            
            if (word.getWord().equals(target)) {
                answer = word.getChangeCnt();
                break;
            }
            
            for (int i = 0; i < words.length; i++) {
                if (!visited[i] && isOneDifferent(word.getWord(), words[i])) {
                    visited[i] = true;
                    queue.add(new ChangedWord(words[i], word.getChangeCnt() + 1));
                }
            }
        }
        
        return answer;
    }
    
    private boolean isOneDifferent(String word, String changedWord) {
        int cnt = 0;
        for (int i = 0; i < word.length(); i++) {
            char oneOfWord = word.charAt(i);
            char oneOfChangedWord = changedWord.charAt(i);
            
            if (oneOfWord != oneOfChangedWord) {
                cnt++;
            }
        }
        
        if (cnt == 1) {
            return true;
        }
        return false;
    }
    
    static class ChangedWord {
        String word;
        int changeCnt;
        
        public ChangedWord(String word, int changeCnt) {
            this.word = word;
            this.changeCnt = changeCnt;
        }
        
        public String getWord() {
            return this.word;
        }
        
        public int getChangeCnt() {
            return this.changeCnt;
        }
    } 
}

File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
from collections import deque

def solution(begin, target, words):
    
    # words 안에 target이 없는 경우 0 반환
    if target not in words:
        return 0
    
    answer = bfs(begin, target, words)
    return answer

def bfs(begin, target, words):
    visited = [0] * len(words)
    q = deque()
    q.append([begin, 0])
    
    while q:
        p = q.popleft()
        word = p[0]
        count = p[1]
        
        if word == target:
            return count
        
        if 0 in visited:
            for i in range(len(words)):
                if visited[i] == 0 and canChange(word, words[i]):
                    visited[i] = 1
                    q.append([words[i], count + 1])
    return 0
  

def canChange(string1, string2):
    diff = 0
    for i in range(len(string1)):
        if string1[i] != string2[i]:
            diff += 1
    return diff == 1